### PR TITLE
feature(ui): show error when proxy process exits

### PR DESCRIPTION
**Pull request by geigerzaehler**
We show the blue screen of death in the app when the proxy process exits. We don’t yet show the reason and additional error information—this is left as a follow-up. See #1095.

### CODE REVIEW

**Review by rudolfs — APPROVED** [review]
Nice refactoring!

Noticed that if you refresh the app when you're on the BSOD, it goes back to the loading screen and then after a minute or so back to BSOD. Not sure this needs to be fixed as refreshing the app is not supposed to happen.
  ref

  Screenshot 2020-11-09 at 17 05 53

Also, I coulnd't test the starting the app when something else has been bound to port 8080, the proxy would just ignore that process and start anyway. Maybe worth opening a new issue?

[proxy-port-binding-weirdness.mp4.zip](https://github.com/radicle-dev/radicle-upstream/files/5511864/proxy-port-binding-weirdness.mp4.zip)

File: native/ipc-types.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Messages send from the main process to the renderer
+export type MainMessage = { type: "ProxyError"; data: ProxyError };
```
[the remark this answers is not in the document]
Comment (xla) [reply native/ipc-types.ts:2]: To flag, we use `type` for tagged enums coming from the proxy API and `kind` in unions in ts.

**Comment by geigerzaehler** [comment]
> Noticed that if you refresh the app while on a BSOD, it goes back to the loading screen and then after a minute or so back to the BSOD. Not sure this needs to be fixed as refreshing the app is not supposed to happen.

I will open a separate issue for this.

> Also, I coulnd't test starting the app while something else has been bound to port 8080, the proxy would just ignore that process and start anyway. Maybe worth opening a new issue?

This is macOS specific behavior. `SO_REUSEPORT` seems to be enabled by default. (See [this article on StackOverflow](https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ))

File: native/main.ts [hunk 2]
```diff
@@ -18,21 +18,87 @@ const proxyPath = path.join(__dirname, "../../proxy");
...
+      event.preventDefault();
+      openExternalLink(url);
+    });
+
+    window.webContents.on("new-window", (event, url) => {
+      event.preventDefault();
+      openExternalLink(url);
+    });
+
+    window.loadURL(`file://${path.join(__dirname, "../public/index.html")}`);
```
[the remark this answers is not in the document]
Comment (geigerzaehler) [reply native/main.ts]: Fixed in https://github.com/radicle-dev/radicle-upstream/pull/1209/commits/641d97cb92336f7163723cad6b39b1d18f9bf90b

File: native/ipc-types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (geigerzaehler) [reply native/ipc-types.ts]: I’ve introduced a `MainMessageKind` enum in https://github.com/radicle-dev/radicle-upstream/pull/1209/commits/641d97cb92336f7163723cad6b39b1d18f9bf90b.

This does not add more type safety though. The TypeScript compiler is smart enough to complain if you make a typo in the string literal. In fact, TypeScript treats these string literals as some kind of “anonymous enum”.

File: native/ipc-types.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+// Payload for the ProxyError `MainMessage`.
+export interface ProxyError {
+  status: number | null;
```
[the remark this answers is not in the document]
Comment (geigerzaehler) [reply native/ipc-types.ts:13]: There’s a difference between the two alternatives. With your change the following code will typecheck whereas it currently does not
~~~typescript
const proxyError: ProxyError = {
  stdout: ""
  stderr: ""
}
~~~
The difference is that we have to provide `status` and `code` explicitly which prevents some mistakes and makes the code easier to understand.

**Review by rudolfs — DISMISSED** [review]

File: native/ipc-types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (FintanH) [reply native/ipc-types.ts]: Ah, good to know :)

File: native/ipc-types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (FintanH) [reply native/ipc-types.ts:13]: Sweet, also good to know :+1:

**Review by rudolfs — APPROVED** [review]

**Review by FintanH — APPROVED** [review]
Using my great frontend skills I can safely say this looks good to me

**Pull request merged**